File: solutions/zig/nucleotide-count/1/nucleotide_count.zig
pub const NucleotideError = error{Invalid};

pub const Counts = struct {
    a: u32,
    c: u32,
    g: u32,
    t: u32,
};

pub fn countNucleotides(s: []const u8) NucleotideError!Counts {
    var counts = Counts{ .a = 0, .c = 0, .g = 0, .t = 0 };
    for (s) |c| switch (c) {
        'A', 'a' => counts.a += 1,
        'C', 'c' => counts.c += 1,
        'G', 'g' => counts.g += 1,
        'T', 't' => counts.t += 1,
        else => return NucleotideError.Invalid,
    };
    return counts;
}

File: solutions/javascript/nucleotide-count/1/nucleotide-count.js
/** @param {string} strand */
export function countNucleotides(strand) {
    const counts = [0, 0, 0, 0];
    for (const nucleotide of strand.toUpperCase()) {
        switch (nucleotide) {
            case 'A':
                counts[0] += 1;
                break;
            case 'C':
                counts[1] += 1;
                break;
            case 'G':
                counts[2] += 1;
                break;
            case 'T':
                counts[3] += 1;
                break;
            default:
                throw new Error('Invalid nucleotide in strand');
        }
    }
    return counts.join(' ');
}
